// @tunein/web-utils/lib/misc/Singleton/Singleton.example.js
import { Singleton } from './Singleton';
class Foo {
    async init(arg) {
        await new Promise((resolve) => setTimeout(resolve, 1));
        return arg;
    }
    doFooThing() {
        console.log('Foo thing');
    }
}
class FooSingleton extends Singleton {
    async init(arg) {
        const foo = new Foo();
        await foo.init(arg);
        this.instance = foo;
        this.#doFooSingletonPrivateThing();
    }
    doFooSingletonThing() {
        console.log('FooSingleton thing');
    }
    #doFooSingletonPrivateThing = () => {
        console.log('FooSingleton private thing');
    };
}
export const foo = new FooSingleton();
foo.init('arg');
await foo.ready;
if (foo.isReady) {
}
foo.doFooSingletonThing();
foo.doFooThing();
foo.init('arg');
class Bar {
    doBarThing() {
        console.log('Bar thing');
    }
}
class BarSingleton extends Singleton {
    init(instance) {
        this.instance = instance;
    }
    doBarSingletonThing() {
        console.log('BarSingleton thing');
    }
}
export const bar = new BarSingleton();
const barInstance = new Bar();
bar.init(barInstance);
bar.init(barInstance);
bar.init(barInstance);
bar.init(barInstance);
bar.doBarSingletonThing();
bar.doBarThing();
